rewrite NIP-46 (#831)
nostr-protocol/nips: 46.md

======

Nostr Connect
-------------

`draft` `optional` `author:tiero` `author:giowe` `author:vforvalerio87` `author:pablof7z` `author:fiatjaf`

This NIP describes a method for 2-way communication between a **remote signer** and a normal Nostr client. The remote signer could be, for example, a hardware device dedicated to signing Nostr events, while the client is a normal Nostr client.

The remote signer generates a connection token in the form

```
<npub1...>#<secret>?relay=wss://...&relay=wss://...
```

The user copies that token and pastes it in the client UI somehow. Then the client can send events of kind `24133` to the specified relays and wait for responses from the remote signer.

The `secret` allows the signer to identify the client's key and, perhaps, assign it certain permissions (e.g. *sign only event `kind:7`, reject encrypt/decrypt requests*)

Event payloads are [NIP-04](04.md)-encrypted JSON blobs that look like JSONRPC.

Events sent by the client to the remote signer have the following format:

```js
{

>   "pubkey": "<client-deterministically-generated-key-for-the-connection>"

Reviewer: wait is it really just for one connection? shouldn't we allow reuse of the client's keys for multiple connections?

Author: Yes, I believe we allow anything.

  "kind": 24133,
  "tags": [
    ["p", "<user-pubkey-hex>"]
  ],
  "content": "nip04_encrypted_json({id: <random-string>, method: <see-below>, params: []})",
  ...
}
```

And the events the remote signer sends to the client have the following format:

```js
  "pubkey": "<user-pubkey>"
  "kind": 24133,
  "tags": [

>     ["p", "<signer-pubkey-hex>"]

Reviewer: "<signer-pubkey-hex>" should be "<client-pubkey-hex>"

Author: I think this is fixed.

  ],
  "content": "nip04_encrypted_json({id: <request-id>, result: <any>, error: <reason-string>})",
  ...
```

### Methods

> - **connect**

Reviewer 2: Is the connect event still needed given the user now pastes the key from the signer into the app?

Author: I agree, but I think implementations out there are relying on this? @pablof7z @staab

Reviewer 3: I don't know, I use pablo's implementation

Reviewer 4: if the user is not sending a secret to apply a pre-authorized list the `connect` method is redundant; nsecBunker doesn't depend on it. Ok with me if we make it optional

Author: Can we just remove it?

Reviewer: I think connect is nice to have for future connections, just so client can verify that it can reach the signer (and if not notify the user). however I am not opposed to removing it entirely. 

If it is not removed, we should at least remove the params (since the signer should be able to identify the client from the event's pubkey already)

  - params: [`pubkey`, `secret`]
  - result: `null`
- **get_public_key**
  - params: []
  - result: `pubkey`
- **sign_event**
  - params: [`event`]
  - result: `event_with_pubkey_id_and_signature`
- **get_relays**
  - params: []
  - result: `{ [url: string]: {read: boolean, write: boolean} }`
- **nip04_encrypt**

>   - params: [`pubkey`, `plaintext`]

Reviewer: for the nip04_encrypt and nip04_decrypt methods below, we should update params to contain `peer pubkey` or some other description that makes it clear it is not the user's own pubkey

Author: I am so confused by this I won't update it until I actually know what I'm doing, but please send a PR if you do.

  - result: `nip4 ciphertext`
- **nip04_decrypt**
  - params: [`pubkey`, `nip4 ciphertext`]

>   - result: [`plaintext`]

Reviewer: since we are revamping this NIP, can we change nip04_decrypt result to not be an array? its the only result that returns an array, was there a reason for that? seems inconsistent.

Author: No, because it would break implementations. This rewrite doesn't try to change the functionality.

Reviewer: ok fine. I guess nip4 will soon be deprecated in favor of nip44 soon anyway.

======

Nostr Connect
-------------

`draft` `optional`

This NIP describes a method for 2-way communication between a **remote signer** and a normal Nostr client. The remote signer could be, for example, a hardware device dedicated to signing Nostr events, while the client is a normal Nostr client.

## Signer Discovery

> The client must somehow be able to contact the signer through a common relay.

Reviewer 5: `The client always starts by generating a random key which is used to communicate with the signer`

Reviewer: no, based on this flow the signer will create the client's random key. also it should be possible (but not required) to reuse this random key for future sessions.

### Started by the signer

The remote signer generates a connection token in the form

```

> <npub1...>#<secret>?relay=wss://...&relay=wss://...

Reviewer 5: Can this also be a URL please? like `nostrconnect://<remote-pubkey>?relay=wss://example.com#<secret>`

this should work in both directions, `metadata` is not required in this direction ofc

Reviewer: I don't think there is any both directions. previously the client created the connection details for the signer. now the signer creates the connection details for the client (so its completely reversed). but maybe i'm misunderstanding...

Author: Do we even need this `secret`? Maybe we should go on the other direction and just paste an `npub1`?

But yes, I personally prefer the URL, the problem is that doing it will break the the nsecBunker flow as it is implemented in everywhere.

Reviewer 5: In bunker the secret is basically a token which will grant certain access to a specific key, it can expire if not claimed and so on, so it has a purpose there.

Its probably because the client is communicating from a random key and this key is not enrolled into the bunker, they claim the token and then the key is associated to those permissions.

Reviewer 4: I think we should go into a URL, clients are the ones parsing the client, so it's not a breaking change for them.

Reviewer 4: How about `nostrconnect://<nip05|pubkey>?relay=...&secret=...`

```

> The user copies that token and pastes it in the client UI somehow. Then the client can send events of kind `24133` to the specified relays and wait for responses from the remote signer.

Reviewer 5: I think the term `<remote-pubkey>` and `<local-pubkey>` are better because the local/remote are reversed depending on which side its initiated from, and ofc `<remote-pubkey>` is always who you're supposed to `p` tag which wasn't clear in this rewrite

Reviewer: only the client can initiate a request and only a signer can send a response. I don't think we need to worry about reverse situations..? maybe i'm not understanding what you are getting at.

right now the `p` tag seems really clear and explicit to me, the remote signer should tag the client pubkey, and the client should tag the remote signer pubkey.

### Started by the client

> The client generates a QR code in the following form (URL-encoded):

Reviewer: ```suggestion
The client generates a random key and creates an URL encoded `nostrconnect` link in the following form (all metadata fields are optional):
```

```

> nostrconnect://<client-key-hex>?relay=wss://...&metadata={"name":"client-name"}

Reviewer 5: Missing the other metadata here? `url` / `description` / `icon`

Author: That wasn't in the NIP before, will add.

> nostrconnect://<client-key-hex>?relay=wss://...&metadata={"name":"...", "url": "...", "description": "..."}

Reviewer: ```suggestion
nostrconnect://<client-key-hex>?relay=wss://...&metadata={"name":"...", "url": "...", "description": "...", "icons": ["...", "..."]}
```

```

> The signer scans the QR code and sends a `connect` message to the client in the specified relays.

Reviewer: ```suggestion
This link is input into the signer using one of the methods below:
- signer scanning a QR code
- deeplinking to the signer app
- pasting it into the signer

Once the link has been entered into the signer, the signer should send an event with payload of { id: "connect", result: `true` } using the indicated relays as acknowledgement that the signer is ready to receive requests from the client.
```
we can change the payload to whatever people think is best. Since the spec says that the payloads for events sent by the signer should look like {id: <request-id>, result: <any>, error: <reason-string>} it was confusing to say that the signer should send a 'connect' message. I figured its easier to just send a dummy initial connection payload. alternatively, we can change the spec to allow signer to send requests as well but I don't like that at all.

Reviewer 6: Signer should be able to send the relay list in this 'connect' message for the app->signer flow. App can't know which relays work best with this signer, app chooses some random relay and puts into nostrconnect: url, signer sends 'connect' back with a list of proper relays as payload. App should then switch and use those relays instead of the random one (or at least add those relays to the pool).

General purpose relays don't work well with nip46 - you quickly hit rate limits for EVENT messages. Plus you can build specialized optimized relays for ephemeral-only load. Plus signer might actually be bundled with a relay to avoid a second network hop. Given all this, it's the signer that should decide which relays work best with it.

Reviewer: I actually completely agree with this (about signer sending back relay with connect response). was going to suggest something similar but thought it would add more complexity and changes so I didn't bother. thank you for throwing it out there!

Reviewer 6: Also, Nostr Wallet Auth proposal https://github.com/nostr-protocol/nips/pull/851/ is solving a similar flow of app->wallet, they too publish a 'connect'-like event after connection is confirmed in the wallet, but they use 30k replaceable event. Their justification is:
> We use a parameterized replaceable event to allow for the wallet and app to
renegotiate the connection if needed and to allow for the wallet being able to look up the connection details later.

Since this 'connect' confirmation event is a new idea in this rewrite maybe we could use a 30k event here too?

Reviewer: I think they use 10k replaceable event, and it doesnt seem to contain information specific to the user (just holds what methods the wallet service supports). 

i'm not sure how comfortable I would be for this nostrConnect connect event to be queryable. 

it is interesting how different these two nips are despite some of their similarities. for example nip47 uses two different kinds for request and response, here we use one.

Reviewer 6: They changed it to 30k, and it's encrypted so it only leaks metadata that a user was connecting to 'some app', so maybe it's not that bad.

Reviewer 6: > it doesnt seem to contain information specific to the user

Their connect event does contain a relay, which is super useful.

## Event payloads

Event payloads are [NIP-04](04.md)-encrypted JSON blobs that look like JSONRPC.

Events sent by the client to the remote signer have the following format:

```js
{
  "pubkey": "<client-key-hex>"
  "kind": 24133,
  "tags": [
    ["p", "<signer-key-hex>"]
  ],
  "content": "nip04_encrypted_json({id: <random-string>, method: <see-below>, params: []})",
  ...
}
```

And the events the remote signer sends to the client have the following format:

```js
  "pubkey": "<signer-key-hex>"
  "kind": 24133,
  "tags": [
    ["p", "<client-key-hex>"]
  ],
  "content": "nip04_encrypted_json({id: <request-id>, result: <any>, error: <reason-string>})",
  ...
```

### Methods

- **connect**

>   - params: [`pubkey`, `secret`]

Reviewer: ```suggestion
  - params: []
```

Author: This is changing the API that is already in use out there in a bunch of places.

Reviewer 2: Since `secret` is being introduced can it please be explained what it is and why? @pablof7z can you help here?

I think this is from:

> if the user is not sending a secret to apply a pre-authorized list the connect method is redundant; nsecBunker doesn't depend on it. Ok with me if we make it optional

What is a pre-authorized list?

Reviewer: @fiatjaf thats true. @rolznz also has a point that @pablof7z said that 'connect' is not needed for nsecbunker, so why are we adding secret here? if secret is needed here for nsecbunker, can we atleast indicate it is optional for client -> signer initializations?

updated my suggestion, also renamed pubkey to indicate that it is the client pubkey. though maybe this is intended to be the signer pubkey? I actually am not sure but its a good thing to state explicitly.

  - result: `null`
- **get_public_key**
  - params: []
  - result: `pubkey`
- **sign_event**
  - params: [`event`]
  - result: `event_with_pubkey_id_and_signature`
- **get_relays**
  - params: []
  - result: `{ [url: string]: {read: boolean, write: boolean} }`
- **nip04_encrypt**

>   - params: [`pubkey`, `plaintext`]

Reviewer: ```suggestion
  - params: [`peer-pubkey`, `plaintext`]
```

  - result: `nip4 ciphertext`
- **nip04_decrypt**

>   - params: [`pubkey`, `nip4 ciphertext`]

Reviewer: ```suggestion
  - params: [`peer-pubkey`, `nip4 ciphertext`]
```
